pageserver/src/pgdatadir_mapping.rs

```diff
@@ -730,11 +730,13 @@ impl Timeline {
     /// Get a KeySpace that covers all the Keys that are in use at the given LSN.
     /// Anything that's not listed maybe removed from the underlying storage (from
     /// that LSN forwards).
+    ///
+    /// The return value is (dense keyspace, sparse keyspace).
     pub(crate) async fn collect_keyspace(
         &self,
         lsn: Lsn,
         ctx: &RequestContext,
-    ) -> Result<KeySpace, CollectKeySpaceError> {
+    ) -> Result<(KeySpace, KeySpace), CollectKeySpaceError> {
```

Comment:
The `(KeySpace, KeySpace)` type is very error-prone if people get them the wrong way around.

Can we make a new `SparseKeySpace` type to avoid that?  It's okay if it's just a simple `struct SparseKeySpace(KeySpace)` for now.

Reply:
fixed in https://github.com/neondatabase/neon/pull/7503/commits/2d87183bd048e7a0b60ffb57dfef23f6f1714d6f